### ISSUE TITLE
A GPU backend with potentially (?) better cross-platform support

### ISSUE THREAD

**Issue by calebwin** [openmm/openmm#2638]
Hello,

@cailinw and I have been working on [an OpenMM fork](https://github.com/cailinw/openmm) where we began trying to create a new OpenMM backend using [emu](https://github.com/calebwin/emu). Emu is a GPGPU library (sort of comparable to OpenCL in level of abstraction) that provides a common abstraction over Vulkan, DX, Metal (and eventually browser through WebAssembly compilation) using SPIR-V as input (which can be generated at compile time from GLSL or other compute shader languages).

I was thinking - and I might be wrong here - that there are some devices (like newer Mac maybe?) that aren't supported by CUDA or OpenCL or don't have as robust support as they have of Vulkan/Metal/DX. Additionally there are mobile/integrated GPUs that maybe aren't supported although I'm not sure if OpenMM intends to target these. If these are issues, I think an Emu-based backend can help with better cross-platform support.

We were mainly motivated by the use of OpenMM in FoldingAtHome where a few people were posting issues which (I think) had to do with lack of GPU support. But we're not 100% sure if an Emu-based backend would help.

Regardless, we began work on the fork and kind of really only got as far as setting up CMake, FFI with Emu (written in Rust), and the `UpdateStateDataKernel`.  (Most of the modified CMake and Rust, C++ is commented pretty well so hopefully the code shouldn't be to hard to parse) We were taking time to understand the GPU-based MD algorithms and school started up again so for now we aren't actively working on it.

That said, I thought I'd post this here in case anyone has thoughts/interested in helping out a bit or if I get some spare time to hack on this.

[the fork we worked on](https://github.com/cailinw/openmm)

**Comment by peastman** [comment]
I'm not familiar with Emu, but we're definitely interested in how to support the widest range of devices.  We just recently created our own (very lightweight) abstraction layer to let us unify a lot of our CUDA and OpenCL code.  See http://docs.openmm.org/development/developerguide/developer.html#common-compute.  In addition to reducing code duplication, a goal of this was to make it easier to support other compute APIs in the feature.  Two of the ones we're most interested in are Metal (if that becomes necessary to support Macs in the future) and HIP (AMD's latest GPGPU language, which is designed to be very similar to CUDA).  Possibly Emu could be supported in the same way?

**Comment by peastman** [comment]
Closing this since it's obsolete.  Emu hasn't been updated in five years.
